### PR TITLE
KAFKA-15045: (KIP-924 pt. 1) New interfaces and stubbed utility classes for pluggable TaskAssignors.

### PR DESCRIPTION
**Pull request by apourchet**
This is the first PR in a sequence to support custom task assignors in Kafka Streams, which was described in KIP 924. It creates and exposes all of the interfaces that will need to be implemented during the refactor of the current task assignment logic.

This PR does not include actual logic, and so has no tests included beyond the typical style and compile checks.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]: These should all be final (with getter APIs but no setters in this class)

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]:
```suggestion
 * Assignment related configs for the Kafka Streams {@link TaskAssignor}.
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]: As mentioned in my comment above, let's take out all the setters in this class

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]: I'd just link to the relevant StreamsConfig, where applicable, instead of trying to rewrite the docs for these:

```suggestion
     * The configured acceptable recovery lag according to 
     * {@link StreamsConfig#ACCEPTABLE_RECOVERY_LAG_CONFIG}
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]: The units of this config is "number of records" not ms. But we can just defer to the config docs and keep it simple here:

```suggestion
     * @return the configured acceptable recovery lag.
```

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]: Or actually, alternatively just get rid of all the `@return` lines in the getter javadocs here, since they pretty much just repeat what we say in the first paragraph of the javadocs

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]:
```suggestion
     * The maximum warmup replicas as configured via
     * {@link StreamsConfig#MAX_WARMUP_REPLICAS_CONFIG}
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]:
```suggestion
     * The number of standby replicas as configured via 
     * {@link StreamsConfig#NUM_STANDBY_REPLICAS_CONFIG}
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]:
```suggestion
     * The probing rebalance interval in milliseconds as configured via 
     * {@link StreamsConfig#PROBING_REBALANCE_INTERVAL_MS}
```

Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]:
```suggestion
     * The probing rebalance interval in milliseconds as configured via 
     * {@link StreamsConfig#PROBING_REBALANCE_INTERVAL_MS_CONFIG}
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]:
```suggestion
     * The rack-aware assignment tags as configured via 
     * {@link StreamsConfig#RACK_AWARE_ASSIGNMENT_TAGS_CONFIG}
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]:
```suggestion
     * The rack-aware assignment traffic cost as configured via 
     * {@link StreamsConfig#RACK_AWARE_ASSIGNMENT_TRAFFIC_COST_CONFIG}
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]:
```suggestion
     * The rack-aware assignment non-overlap cost as configured via 
     * {@link StreamsConfig#RACK_AWARE_ASSIGNMENT_NON_OVERLAP_COST_CONFIG}
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsAssignment.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsAssignment.java]:
```suggestion
     * @return the followup rebalance deadline in epoch time, after which this KafkaStreams client will trigger a new rebalance
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsAssignment.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsAssignment.java]: Remove this, doesn't add much and it's one of those things that we'll almost definitely forget to update if we ever change this class

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsAssignment.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsAssignment.java]: Same here: remove this. The code is pretty self-explanatory and if we ever added a new task type there's a good chance we could forget to update the comment. 

This isn't strictly speaking a style guideline, but it's a general philosophy/rule of thumb for comments/docs in Kafka/Kafka Streams. Since so many different people are updating the code over such a long period of time, docs easily end up out of date and become misleading if not outright incorrect. Doesn't mean you shouldn't include docs, but it's good to consider whether they're adding enough to justify any risk of going out of date

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsState.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsState.java:62]: I realize I forgot this in the KIP, but we should throw an exception if this method is called and the user did not request task lags be computed. Not sure what exception type makes the most sense here, maybe  IllegalArgumentException or UnsupportedOperationException?

Whichever you choose, just add a `@throws` statement to the javadocs here

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsState.java:62]: ditto for the `#prevTasksByLag` and `#statefulTasksToLagSums` methods below

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/ProcessID.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/ProcessID.java]: nit: rename to `ProcessId` (lowercase 'd') to be more consistent with naming in Streams

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentException.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentException.java]: I think TaskAssignmentException already exists and is part of the public API, so we don't need to add it. All the exceptions for Streams are under `org.apache.kafka.streams.errors`

(btw: if you ever actually do need to introduce a new exception type, it should probably extend StreamsException. Kafka Streams exceptions all extend StreamsException, things in core or the clients will extend KafkaException or one of its subtypes, it's a whole thing)

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]:
```suggestion
 * A set of utilities to help implement task assignment via the {@link TaskAssignor}
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignor.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignor.java:30]: I do think it's worth expanding the javadocs here, potentially even giving an inline example like we do in some of the major classes in streams and the clients, but it probably makes sense to wait until we're at the end in case anything changes along the way. Since as so many of my comments bring up, we often forget to update javadocs lol

**Review by ableegoldman — COMMENTED** [review]
Left a number of mostly cosmetic suggestions, but on the whole it looks good. Should be ready to merge once you address all the comments

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsState.java
Comment (apourchet) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsState.java:62]: IllegalArgumentException usually refers to an argument that was passed in, as opposed to some global misconfiguration, so I'll go for UnsupportedOperationException unless someone else feels strongly about it.

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/ApplicationState.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/ApplicationState.java]:
```suggestion
 * A read-only metadata class representing the state of the application and the current rebalance.
 * This class wraps all the input parameters to the task assignment, including the current state
 * of each KafkaStreams client with at least one StreamThread participating in this rebalance, the
 * assignment-related configs, and the tasks to be assigned. 
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]: nit: keep the "rack-aware" prefix for all the rack-aware related configs. So:
1. `trafficCost --> `rackAwareTrafficCost`
2. `nonOverlapCost` -> `rackAwareNonOverlapCost`
3. `assignmentStrategy` --> `rackAwareAssignmentStrategy`

(and update the getter method names accordingly)

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]: nit: fix alignment for this and below params

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]: nit: just call the below constructor

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/AssignmentConfigs.java]: I'd make all of these primitive types (ie Long --> long, Integer --> int). Otherwise it seems like they can be null (and they can't)

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsState.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/KafkaStreamsState.java:62]: Makes sense. We definitely abuse IllegalArgumentException to mean more than that elsewhere in Streams but no reason to perpetuate that bad habit lol

**Review by ableegoldman — APPROVED** [review]
LGTM!

**Comment by ableegoldman** [comment]
Test failures were unrelated. Merged to trunk

**Pull request merged**
